For this project we have a modified dataset of German ebay website of cars. We will perform various tasks such as  analyzing, cleaning, modifying, comparing and gets information about whole dataframe.

Lets first import pandas and numpy and import the csv file into a variable called a.

Due to some encoding errors you might expect to see unicode tag such as 'u' at the begining of every elements of a column, if you see just ignore it as it doesn't affect any of the operation we will perform throughout the project.

In [1]:
import pandas as pd
import numpy as np

In [2]:
a = pd.read_csv('autos.csv', encoding = 'latin1')

Now lets dive in to get some information about the dataframe. We will use Dataframe.info, Dataframe.head(), Dataframe.columns

In [3]:
a.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

Dataframe.info() showing us that we have 50000 rows and 20 columns, some of the columns got null objects and the whole Dataframe is about 8.5 MB.

In [4]:
a.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# Lets Dive in throughly through the DataFrame and make some changes

By executing Dataframe.head() we can see the first five rows of all columns. At first glace we can see that the seller, offerType, vehicleType, gearbox, notRepairedDamage column got some German words in it that need to be translated.

price and odometer columns got the value as a string format which we will change it to numeric.

All coumns are in camelcase format and some of them are quite lengthy which is quite hard to work with.
So we will change all columns name from camelcase to snakecase first and then shorten some of the columns name 


In [5]:
a.columns

Index([u'dateCrawled', u'name', u'seller', u'offerType', u'price', u'abtest',
       u'vehicleType', u'yearOfRegistration', u'gearbox', u'powerPS', u'model',
       u'odometer', u'monthOfRegistration', u'fuelType', u'brand',
       u'notRepairedDamage', u'dateCreated', u'nrOfPictures', u'postalCode',
       u'lastSeen'],
      dtype='object')

In [6]:
new_columns = []

In [7]:
def change_case(str): 
    res = [str[0].lower()] 
    for c in str[1:]: 
        if c in ('ABCDEFGHIJKLMNOPQRSTUVWXYZ'): 
            res.append('_') 
            res.append(c.lower()) 
        else: 
            res.append(c) 
      
    return ''.join(res)


In [8]:
for i in a.columns:
    b = change_case(i)
    new_columns.append(b)

In [9]:
a.columns = new_columns

In [10]:
a.rename(columns ={'year_of_registration':'registration_year', 'month_of_registration':'registration_month', 'odometer':'odometer_km', 'not_repaired_damage':'unrepaired_damage', 'date_created':'ad_created'}, inplace = True)

In [11]:
a.columns

Index([      u'date_crawled',               u'name',             u'seller',
               u'offer_type',              u'price',             u'abtest',
             u'vehicle_type',  u'registration_year',            u'gearbox',
                u'power_p_s',              u'model',        u'odometer_km',
       u'registration_month',          u'fuel_type',              u'brand',
        u'unrepaired_damage',         u'ad_created',     u'nr_of_pictures',
              u'postal_code',          u'last_seen'],
      dtype='object')

After making change to the columns name, lets call DataFrame.head() to see how the change affects.

In [12]:
a.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model odometer_km  registration_month fuel_type  \
0    manuell        158  andere   150,000km                   3       lpg   
1  automatik        286     7er   150,000km                   6    benzin   
2    manuell        102    golf    70,000km                   7    benzin   
3  automatik         71  fortwo    70,000km                   6    benzin   
4    manuell          0   focus   150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

nr_of_pictures column got 0 throuhout the dataset, thats why its likely to be dropped
registration_year needs to be looked further down
price and odometer_km stored the value as strings, so that need to be converted to int

In [13]:
a.describe()

registration_year     power_p_s  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000           50000   
mean         2005.073280    116.355920            5.723360               0   
std           105.712813    209.216627            3.711984               0   
min          1000.000000      0.000000            0.000000               0   
25%          1999.000000     70.000000            3.000000               0   
50%          2003.000000    105.000000            6.000000               0   
75%          2008.000000    150.000000            9.000000               0   
max          9999.000000  17700.000000           12.000000               0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [14]:
a['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

In [15]:
a.drop(['nr_of_pictures'], axis = 1, inplace = True)

# Lets change the price and odometer_km column dtypes from string to int

Now lets dig in to the price and odometer_km columns. As we can see that both the columns got the value stored as strings that need to be change to numeric value.

In [16]:
a.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model odometer_km  registration_month fuel_type  \
0    manuell        158  andere   150,000km                   3       lpg   
1  automatik        286     7er   150,000km                   6    benzin   
2    manuell        102    golf    70,000km                   7    benzin   
3  automatik         71  fortwo    70,000km                   6    benzin   
4    manuell          0   focus   150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00        79588   
1         bmw              nein  2016-04-04 00:00:00        71034   
2  volkswagen              nein  2016-03-26 00:00:00        35394   
3       smart              nein  2016-03-12 00:00:00        33729   
4        ford              nein  2016-04-01 00:00:00        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [17]:
a['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [18]:
a['price'] = a['price'].str.replace('$', '')
a['price'] = a['price'].str.replace(',', '')
a['price'] = a['price'].astype(int)

In [19]:
a['odometer_km'] = a['odometer_km'].str.replace(',', '')
a['odometer_km'] = a['odometer_km'].str.replace('km', '')
a['odometer_km'] = a['odometer_km'].astype(int)

In [20]:
a['price'].describe()

count       50000.00000
mean         9840.04376
std        481104.38050
min             0.00000
25%          1100.00000
50%          2950.00000
75%          7200.00000
max      99999999.00000
Name: price, dtype: float64

In [21]:
a['price'].value_counts().sort_index(ascending = False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

In [22]:
a['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [23]:
a['odometer_km'].value_counts().sort_index(ascending = False).head(10)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
Name: odometer_km, dtype: int64

# Remmoving Outliers from price column

price column got a way too expensive and unrealistic value of 0 of used cars so this columns need to adjusted a bit, anything between 1000 and 300000 will be removed.

150000 km for a used car still okay same goes for 5000 km, so odometer_km is ok.

In [24]:
a = a[a['price'].between(1000, 300000)]

In [25]:
a['price'].describe()

count     38624.000000
mean       7237.757974
std        9384.523689
min        1000.000000
25%        2200.000000
50%        4350.000000
75%        8950.000000
max      299000.000000
Name: price, dtype: float64

# Sorting the date columns

All the date columns got the same format first 10 characters are the date format and the rest are timestamps, so lets extract the date value and remove the timestamp

In [26]:
a.head(10)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   

    seller offer_type  price   abtest vehicle_type  registration_year  \
0   privat    Angebot   5000  control          bus               2004   
1   privat    Angebot   8500  control    limousine               1997   
2   privat    Angebot   8990     test    limousine               2009   
3   privat    Angebot   4350  control   kleinwagen               2007   
4   privat    Angebot   1350     test        kombi               2003   
5   privat    Angebot   7900     test          bus               2006   
7   privat    Angebot   1990  control    limousine               1998   
12  privat    Angebot   5299  control   kleinwagen               2010   
13  privat    Angebot   1350  control    limousine               1999   
14  privat    Angebot   3999     test   kleinwagen               2007   

      gearbox  power_p_s    model  odometer_km  registration_month fuel_type  \
0     manuell        158   andere       150000                   3       lpg   
1   automatik        286      7er       150000                   6    benzin   
2     manuell        102     golf        70000                   7    benzin   
3   automatik         71   fortwo        70000                   6    benzin   
4     manuell          0    focus       150000                   7    benzin   
5   automatik        150  voyager       150000                   4    diesel   
7     manuell         90     golf       150000                  12    diesel   
12  automatik         71   fortwo        50000                   9    benzin   
13    manuell        101       a3       150000                  11    benzin   
14    manuell         75     clio       150000                   9    benzin   

         brand unrepaired_damage           ad_created  postal_code  \
0      peugeot              nein  2016-03-26 00:00:00        79588   
1          bmw              nein  2016-04-04 00:00:00        71034   
2   volkswagen              nein  2016-03-26 00:00:00        35394   
3        smart              nein  2016-03-12 00:00:00        33729   
4         ford              nein  2016-04-01 00:00:00        39218   
5     chrysler               NaN  2016-03-21 00:00:00        22962   
7   volkswagen              nein  2016-03-16 00:00:00        53474   
12       smart              nein  2016-03-31 00:00:00        34590   
13        audi              nein  2016-03-23 00:00:00        12043   
14     renault               NaN  2016-03-23 00:00:00        81737   

              last_seen  
0   2016-04-06 06:45:54  
1   2016-04-06 14:45:08  
2   2016-04-06 20:15:37  
3   2016-03-15 03:16:28  
4   2016-04-01 14:38:50  
5   2016-04-06 09:45:21  
7   2016-04-07 03:17:32  
12  2016-04-06 14:17:52  
13  2016-04-01 14:17:13  
14  2016-04-01 15:46:47

In [27]:
a['date_crawled'].describe()

count                   38624
unique                  37556
top       2016-03-11 22:38:16
freq                        3
Name: date_crawled, dtype: object

In [28]:
a[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [29]:
a['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2016-03-05    0.025554
2016-03-06    0.013877
2016-03-07    0.035134
2016-03-08    0.032622
2016-03-09    0.032467
2016-03-10    0.033321
2016-03-11    0.032803
2016-03-12    0.037386
2016-03-13    0.016000
2016-03-14    0.036635
2016-03-15    0.033632
2016-03-16    0.029075
2016-03-17    0.030473
2016-03-18    0.012842
2016-03-19    0.035134
2016-03-20    0.038163
2016-03-21    0.037283
2016-03-22    0.032519
2016-03-23    0.032208
2016-03-24    0.029023
2016-03-25    0.030525
2016-03-26    0.033114
2016-03-27    0.031379
2016-03-28    0.035367
2016-03-29    0.033969
2016-03-30    0.033036
2016-03-31    0.031405
2016-04-01    0.034616
2016-04-02    0.036299
2016-04-03    0.039147
2016-04-04    0.036868
2016-04-05    0.013360
2016-04-06    0.003262
2016-04-07    0.001502
Name: date_crawled, dtype: float64

In [30]:
a['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2015-06-11    0.000026
2015-08-10    0.000026
2015-09-09    0.000026
2015-11-10    0.000026
2015-12-30    0.000026
2016-01-03    0.000026
2016-01-07    0.000026
2016-01-10    0.000052
2016-01-13    0.000026
2016-01-14    0.000026
2016-01-16    0.000026
2016-01-22    0.000026
2016-01-27    0.000078
2016-01-29    0.000026
2016-02-01    0.000026
2016-02-02    0.000052
2016-02-05    0.000052
2016-02-07    0.000026
2016-02-09    0.000026
2016-02-11    0.000026
2016-02-12    0.000052
2016-02-14    0.000052
2016-02-16    0.000026
2016-02-17    0.000026
2016-02-18    0.000052
2016-02-19    0.000078
2016-02-20    0.000026
2016-02-21    0.000052
2016-02-22    0.000026
2016-02-23    0.000104
                ...   
2016-03-09    0.032648
2016-03-10    0.032985
2016-03-11    0.033062
2016-03-12    0.037127
2016-03-13    0.017657
2016-03-14    0.034978
2016-03-15    0.033451
2016-03-16    0.029645
2016-03-17    0.030163
2016-03-18    0.013282
2016-03-19    0.034072
2016-03-20    0.038266
2016-03-21 

In [31]:
a['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2016-03-05    0.001087
2016-03-06    0.003573
2016-03-07    0.004557
2016-03-08    0.006240
2016-03-09    0.008906
2016-03-10    0.009813
2016-03-11    0.011728
2016-03-12    0.022188
2016-03-13    0.008389
2016-03-14    0.011987
2016-03-15    0.014991
2016-03-16    0.015457
2016-03-17    0.026383
2016-03-18    0.007379
2016-03-19    0.014602
2016-03-20    0.019806
2016-03-21    0.019677
2016-03-22    0.020790
2016-03-23    0.017916
2016-03-24    0.018538
2016-03-25    0.017761
2016-03-26    0.016078
2016-03-27    0.014059
2016-03-28    0.019418
2016-03-29    0.020764
2016-03-30    0.023457
2016-03-31    0.022732
2016-04-01    0.023198
2016-04-02    0.024907
2016-04-03    0.024441
2016-04-04    0.023379
2016-04-05    0.131136
2016-04-06    0.234699
2016-04-07    0.139965
Name: last_seen, dtype: float64

In [32]:
a[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

# Removing outliers from the registration_year column

In [33]:
a['registration_year'].describe()

count    38624.000000
mean      2005.679370
std         86.687361
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

This column got some unrealistic year value like 1000 and 9999, so this column need to be cleaned up a bit, as car was invented around 1900 so anything beyond 1900 and 2020 will be removed.

In [34]:
a = a[a['registration_year'].between(1990, 2020)]

In [35]:
a['registration_year'].describe()

count    37511.000000
mean      2005.206926
std          5.698639
min       1990.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       2019.000000
Name: registration_year, dtype: float64

# Aggregate the top brand with mean price

We will extract the most common cars and set it as top brand and then will calculate the mean price by aggregating by the brand with price. After that we will figure out the mean milege and then will construct both the mean_price and mean_milege as a pandas series, finally we will construct those series as DataFrame.

In [36]:
top_brand = a['brand'].value_counts().head(20).index

In [37]:
top_brand

Index([u'volkswagen', u'bmw', u'mercedes_benz', u'audi', u'opel', u'ford',
       u'renault', u'peugeot', u'fiat', u'skoda', u'seat', u'smart', u'mazda',
       u'toyota', u'citroen', u'nissan', u'hyundai', u'mini', u'volvo',
       u'kia'],
      dtype='object')

In [38]:
mean_price = {}

In [39]:
for i in top_brand:
    d = a[a['brand'] == i]
    price_brand = d['price']
    mean_price[i] = price_brand.mean()

In [40]:
mean_price

{u'audi': 10254.823132919924,
 u'bmw': 9092.9437460046884,
 u'citroen': 4445.157196969697,
 u'fiat': 3849.0414052697615,
 u'ford': 4893.0310360565954,
 u'hyundai': 6174.4830917874397,
 u'kia': 6743.6195286195289,
 u'mazda': 5223.9312839059676,
 u'mercedes_benz': 9159.6329708353933,
 u'mini': 10667.012165450122,
 u'nissan': 6336.4827586206893,
 u'opel': 4161.1965862271927,
 u'peugeot': 3914.6227657572908,
 u'renault': 3461.3150589868146,
 u'seat': 5504.8278335724535,
 u'skoda': 6834.0820584144649,
 u'smart': 3718.3719135802471,
 u'toyota': 5537.7359855334535,
 u'volkswagen': 6556.064059170114,
 u'volvo': 5985.5723076923077}

In [41]:
price_series = pd.Series(mean_price)

In [42]:
price_milege_df = pd.DataFrame(price_series, columns = ['mean_price'])

In [43]:
price_milege_df

mean_price
audi           10254.823133
bmw             9092.943746
citroen         4445.157197
fiat            3849.041405
ford            4893.031036
hyundai         6174.483092
kia             6743.619529
mazda           5223.931284
mercedes_benz   9159.632971
mini           10667.012165
nissan          6336.482759
opel            4161.196586
peugeot         3914.622766
renault         3461.315059
seat            5504.827834
skoda           6834.082058
smart           3718.371914
toyota          5537.735986
volkswagen      6556.064059
volvo           5985.572308

In [44]:
milege = {}

In [45]:
for k in b:
    l = a[a['brand'] == k]
    mean_milege = l['odometer_km']
    milege[k] = mean_milege.mean()

In [46]:
milege_series = pd.Series(milege)

In [47]:
price_milege_df['mean_milege'] = milege_series

In [94]:
price_milege_df

mean_price  mean_milege
audi           10254.823133          NaN
bmw             9092.943746          NaN
citroen         4445.157197          NaN
fiat            3849.041405          NaN
ford            4893.031036          NaN
hyundai         6174.483092          NaN
kia             6743.619529          NaN
mazda           5223.931284          NaN
mercedes_benz   9159.632971          NaN
mini           10667.012165          NaN
nissan          6336.482759          NaN
opel            4161.196586          NaN
peugeot         3914.622766          NaN
renault         3461.315059          NaN
seat            5504.827834          NaN
skoda           6834.082058          NaN
smart           3718.371914          NaN
toyota          5537.735986          NaN
volkswagen      6556.064059          NaN
volvo           5985.572308          NaN

# Filtering and translating the German words to English

Some of the columns such as seller, offer_type, vehicle_type, gearbox, undamaged_repair have value in German which will be changed to English

In [49]:
a.head(10)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   

    seller offer_type  price   abtest vehicle_type  registration_year  \
0   privat    Angebot   5000  control          bus               2004   
1   privat    Angebot   8500  control    limousine               1997   
2   privat    Angebot   8990     test    limousine               2009   
3   privat    Angebot   4350  control   kleinwagen               2007   
4   privat    Angebot   1350     test        kombi               2003   
5   privat    Angebot   7900     test          bus               2006   
7   privat    Angebot   1990  control    limousine               1998   
12  privat    Angebot   5299  control   kleinwagen               2010   
13  privat    Angebot   1350  control    limousine               1999   
14  privat    Angebot   3999     test   kleinwagen               2007   

      gearbox  power_p_s    model  odometer_km  registration_month fuel_type  \
0     manuell        158   andere       150000                   3       lpg   
1   automatik        286      7er       150000                   6    benzin   
2     manuell        102     golf        70000                   7    benzin   
3   automatik         71   fortwo        70000                   6    benzin   
4     manuell          0    focus       150000                   7    benzin   
5   automatik        150  voyager       150000                   4    diesel   
7     manuell         90     golf       150000                  12    diesel   
12  automatik         71   fortwo        50000                   9    benzin   
13    manuell        101       a3       150000                  11    benzin   
14    manuell         75     clio       150000                   9    benzin   

         brand unrepaired_damage           ad_created  postal_code  \
0      peugeot              nein  2016-03-26 00:00:00        79588   
1          bmw              nein  2016-04-04 00:00:00        71034   
2   volkswagen              nein  2016-03-26 00:00:00        35394   
3        smart              nein  2016-03-12 00:00:00        33729   
4         ford              nein  2016-04-01 00:00:00        39218   
5     chrysler               NaN  2016-03-21 00:00:00        22962   
7   volkswagen              nein  2016-03-16 00:00:00        53474   
12       smart              nein  2016-03-31 00:00:00        34590   
13        audi              nein  2016-03-23 00:00:00        12043   
14     renault               NaN  2016-03-23 00:00:00        81737   

              last_seen  
0   2016-04-06 06:45:54  
1   2016-04-06 14:45:08  
2   2016-04-06 20:15:37  
3   2016-03-15 03:16:28  
4   2016-04-01 14:38:50  
5   2016-04-06 09:45:21  
7   2016-04-07 03:17:32  
12  2016-04-06 14:17:52  
13  2016-04-01 14:17:13  
14  2016-04-01 15:46:47

In [50]:
a['seller'].value_counts()

privat    37511
Name: seller, dtype: int64

In [51]:
a['offer_type'].value_counts()

Angebot    37511
Name: offer_type, dtype: int64

In [52]:
a['vehicle_type'].value_counts()

limousine     10171
kombi          7496
kleinwagen     6766
bus            3639
cabrio         2665
coupe          1969
suv            1865
andere          268
Name: vehicle_type, dtype: int64

In [53]:
a['gearbox'].value_counts()

manuell      27082
automatik     9109
Name: gearbox, dtype: int64

In [54]:
a['unrepaired_damage'].value_counts()

nein    29730
ja       2300
Name: unrepaired_damage, dtype: int64

Now lets change those words to English

In [55]:
a['seller'] = a['seller'].map({'privat':'private'})

In [56]:
a['offer_type'] = a['offer_type'].map({'Angebot':'offer'})

In [57]:
a['vehicle_type']= a['vehicle_type'].map({'limousine':'limousine', 'kombi':'van', 'kleinwagen':'supermini', 'bus':'bus', 'cabrio':'convertible', 'coupe':'coupe', 'suv':'suv', 'andere':'other'})

In [58]:
a['gearbox'] = a['gearbox'].map({'manuell':'manual', 'automatik':'automatic'})

In [59]:
a['unrepaired_damage'] = a['unrepaired_damage'].map({'nein':'no', 'ja':'yes'})

In [60]:
a.head(10)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   

     seller offer_type  price   abtest vehicle_type  registration_year  \
0   private      offer   5000  control          bus               2004   
1   private      offer   8500  control    limousine               1997   
2   private      offer   8990     test    limousine               2009   
3   private      offer   4350  control    supermini               2007   
4   private      offer   1350     test          van               2003   
5   private      offer   7900     test          bus               2006   
7   private      offer   1990  control    limousine               1998   
12  private      offer   5299  control    supermini               2010   
13  private      offer   1350  control    limousine               1999   
14  private      offer   3999     test    supermini               2007   

      gearbox  power_p_s    model  odometer_km  registration_month fuel_type  \
0      manual        158   andere       150000                   3       lpg   
1   automatic        286      7er       150000                   6    benzin   
2      manual        102     golf        70000                   7    benzin   
3   automatic         71   fortwo        70000                   6    benzin   
4      manual          0    focus       150000                   7    benzin   
5   automatic        150  voyager       150000                   4    diesel   
7      manual         90     golf       150000                  12    diesel   
12  automatic         71   fortwo        50000                   9    benzin   
13     manual        101       a3       150000                  11    benzin   
14     manual         75     clio       150000                   9    benzin   

         brand unrepaired_damage           ad_created  postal_code  \
0      peugeot                no  2016-03-26 00:00:00        79588   
1          bmw                no  2016-04-04 00:00:00        71034   
2   volkswagen                no  2016-03-26 00:00:00        35394   
3        smart                no  2016-03-12 00:00:00        33729   
4         ford                no  2016-04-01 00:00:00        39218   
5     chrysler               NaN  2016-03-21 00:00:00        22962   
7   volkswagen                no  2016-03-16 00:00:00        53474   
12       smart                no  2016-03-31 00:00:00        34590   
13        audi                no  2016-03-23 00:00:00        12043   
14     renault               NaN  2016-03-23 00:00:00        81737   

              last_seen  
0   2016-04-06 06:45:54  
1   2016-04-06 14:45:08  
2   2016-04-06 20:15:37  
3   2016-03-15 03:16:28  
4   2016-04-01 14:38:50  
5   2016-04-06 09:45:21  
7   2016-04-07 03:17:32  
12  2016-04-06 14:17:52  
13  2016-04-01 14:17:13  
14  2016-04-01 15:46:47

# Now lets change the date value from string to numeric

Looking at the date columns such as date_crawled, ad_created, last_seen are in string format, we will convert it to numeric value.

In [61]:
a.head(10)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   

     seller offer_type  price   abtest vehicle_type  registration_year  \
0   private      offer   5000  control          bus               2004   
1   private      offer   8500  control    limousine               1997   
2   private      offer   8990     test    limousine               2009   
3   private      offer   4350  control    supermini               2007   
4   private      offer   1350     test          van               2003   
5   private      offer   7900     test          bus               2006   
7   private      offer   1990  control    limousine               1998   
12  private      offer   5299  control    supermini               2010   
13  private      offer   1350  control    limousine               1999   
14  private      offer   3999     test    supermini               2007   

      gearbox  power_p_s    model  odometer_km  registration_month fuel_type  \
0      manual        158   andere       150000                   3       lpg   
1   automatic        286      7er       150000                   6    benzin   
2      manual        102     golf        70000                   7    benzin   
3   automatic         71   fortwo        70000                   6    benzin   
4      manual          0    focus       150000                   7    benzin   
5   automatic        150  voyager       150000                   4    diesel   
7      manual         90     golf       150000                  12    diesel   
12  automatic         71   fortwo        50000                   9    benzin   
13     manual        101       a3       150000                  11    benzin   
14     manual         75     clio       150000                   9    benzin   

         brand unrepaired_damage           ad_created  postal_code  \
0      peugeot                no  2016-03-26 00:00:00        79588   
1          bmw                no  2016-04-04 00:00:00        71034   
2   volkswagen                no  2016-03-26 00:00:00        35394   
3        smart                no  2016-03-12 00:00:00        33729   
4         ford                no  2016-04-01 00:00:00        39218   
5     chrysler               NaN  2016-03-21 00:00:00        22962   
7   volkswagen                no  2016-03-16 00:00:00        53474   
12       smart                no  2016-03-31 00:00:00        34590   
13        audi                no  2016-03-23 00:00:00        12043   
14     renault               NaN  2016-03-23 00:00:00        81737   

              last_seen  
0   2016-04-06 06:45:54  
1   2016-04-06 14:45:08  
2   2016-04-06 20:15:37  
3   2016-03-15 03:16:28  
4   2016-04-01 14:38:50  
5   2016-04-06 09:45:21  
7   2016-04-07 03:17:32  
12  2016-04-06 14:17:52  
13  2016-04-01 14:17:13  
14  2016-04-01 15:46:47

In [62]:
a['date_crawled'] = a['date_crawled'].str[:10].str.replace('-', '').astype(int)

In [63]:
a['ad_created'] = a['ad_created'].str[:10].str.replace('-', '').astype(int)

In [64]:
a['last_seen'] = a['last_seen'].str[:10].str.replace('-', '').astype(int)

In [65]:
a.head(10)

date_crawled                                               name   seller  \
0       20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1       20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2       20160326                         Volkswagen_Golf_1.6_United  private   
3       20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4       20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   
5       20160321  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...  private   
7       20160316                               Golf_IV_1.9_TDI_90PS  private   
12      20160331  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...  private   
13      20160323                                 Audi_A3_1.6_tuning  private   
14      20160323  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...  private   

   offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0       offer   5000  control          bus               2004     manual   
1       offer   8500  control    limousine               1997  automatic   
2       offer   8990     test    limousine               2009     manual   
3       offer   4350  control    supermini               2007  automatic   
4       offer   1350     test          van               2003     manual   
5       offer   7900     test          bus               2006  automatic   
7       offer   1990  control    limousine               1998     manual   
12      offer   5299  control    supermini               2010  automatic   
13      offer   1350  control    limousine               1999     manual   
14      offer   3999     test    supermini               2007     manual   

    power_p_s    model  odometer_km  registration_month fuel_type       brand  \
0         158   andere       150000                   3       lpg     peugeot   
1         286      7er       150000                   6    benzin         bmw   
2         102     golf        70000                   7    benzin  volkswagen   
3          71   fortwo        70000                   6    benzin       smart   
4           0    focus       150000                   7    benzin        ford   
5         150  voyager       150000                   4    diesel    chrysler   
7          90     golf       150000                  12    diesel  volkswagen   
12         71   fortwo        50000                   9    benzin       smart   
13        101       a3       150000                  11    benzin        audi   
14         75     clio       150000                   9    benzin     renault   

   unrepaired_damage  ad_created  postal_code  last_seen  
0                 no    20160326        79588   20160406  
1                 no    20160404        71034   20160406  
2                 no    20160326        35394   20160406  
3                 no    20160312        33729   20160315  
4                 no    20160401        39218   20160401  
5                NaN    20160321        22962   20160406  
7                 no    20160316        53474   20160407  
12                no    20160331        34590   20160406  
13                no    20160323        12043   20160401  
14               NaN    20160323        81737   20160401

# Finding the most common make/model combinations

Next, We will figure out the most common make/model comnbinations

In [66]:
top_brand = a['brand'].value_counts().head(5)

In [67]:
top_brand

volkswagen       7977
bmw              4693
mercedes_benz    4046
audi             3709
opel             3398
Name: brand, dtype: int64

In [68]:
model_freq_vw = {}

In [69]:
import operator
vw = a['brand'] == 'volkswagen'
vw_models = a[vw]['model']

model_freq_vw = {}

for m in vw_models:
    if m not in model_freq_vw:
        model_freq_vw[m] = 1
    else:
        model_freq_vw[m] += 1
sorted_vw = sorted(model_freq_vw.items(), key=operator.itemgetter(1), reverse=True)
sorted_vw[0]

(u'golf', 3000)

In [70]:
def brand_model_freq(b):
    
    brand_filter = a['brand'] == b
    brand_models = a[brand_filter]['model']
    
    model_freq = {}
    
    for m in brand_models:
        if m not in model_freq:
            model_freq[m] = 1
        else:
            model_freq[m] += 1
    
    sorted_brand = sorted(model_freq.items(), key=operator.itemgetter(1), reverse=True)
    
    return sorted_brand[0]

In [71]:
audi = brand_model_freq('audi')

In [72]:
opel = brand_model_freq('opel')

In [73]:
bmw = brand_model_freq('bmw')

In [74]:
mercedes_benz = brand_model_freq('mercedes_benz')

In [75]:
volkswagen = brand_model_freq('volkswagen')

In [76]:
audi

(u'a4', 1110)

In [77]:
opel

(u'astra', 1019)

In [78]:
bmw

(u'3er', 2297)

In [79]:
mercedes_benz

(u'c_klasse', 1057)

In [80]:
volkswagen

(u'golf', 3000)

# Lets find out if price change with milege

We are going to Aggregate average price with odometer_km to see if the price change with mileges.

In [81]:
group_a = a.loc[a['odometer_km'] == 150000.0, 'price']

In [82]:
group_b = a.loc[a['odometer_km'].between(100000, 149999), 'price']

In [83]:
group_c = a.loc[a['odometer_km'] < 100000, 'price']

In [84]:
group_a.mean()

4766.8076670029386

In [85]:
group_b.mean()

7528.4986706547024

In [86]:
group_c.mean()

13202.11249137336

# Does price depends on repaired and unrepaired damage cars

We are going to find out the price difference between repaired and unrepaired cars, in that way we can say which one is expensive or cheap.

In [87]:
a.head(10)

date_crawled                                               name   seller  \
0       20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1       20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2       20160326                         Volkswagen_Golf_1.6_United  private   
3       20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4       20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   
5       20160321  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...  private   
7       20160316                               Golf_IV_1.9_TDI_90PS  private   
12      20160331  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...  private   
13      20160323                                 Audi_A3_1.6_tuning  private   
14      20160323  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...  private   

   offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0       offer   5000  control          bus               2004     manual   
1       offer   8500  control    limousine               1997  automatic   
2       offer   8990     test    limousine               2009     manual   
3       offer   4350  control    supermini               2007  automatic   
4       offer   1350     test          van               2003     manual   
5       offer   7900     test          bus               2006  automatic   
7       offer   1990  control    limousine               1998     manual   
12      offer   5299  control    supermini               2010  automatic   
13      offer   1350  control    limousine               1999     manual   
14      offer   3999     test    supermini               2007     manual   

    power_p_s    model  odometer_km  registration_month fuel_type       brand  \
0         158   andere       150000                   3       lpg     peugeot   
1         286      7er       150000                   6    benzin         bmw   
2         102     golf        70000                   7    benzin  volkswagen   
3          71   fortwo        70000                   6    benzin       smart   
4           0    focus       150000                   7    benzin        ford   
5         150  voyager       150000                   4    diesel    chrysler   
7          90     golf       150000                  12    diesel  volkswagen   
12         71   fortwo        50000                   9    benzin       smart   
13        101       a3       150000                  11    benzin        audi   
14         75     clio       150000                   9    benzin     renault   

   unrepaired_damage  ad_created  postal_code  last_seen  
0                 no    20160326        79588   20160406  
1                 no    20160404        71034   20160406  
2                 no    20160326        35394   20160406  
3                 no    20160312        33729   20160315  
4                 no    20160401        39218   20160401  
5                NaN    20160321        22962   20160406  
7                 no    20160316        53474   20160407  
12                no    20160331        34590   20160406  
13                no    20160323        12043   20160401  
14               NaN    20160323        81737   20160401

In [88]:
unrepaired = a['unrepaired_damage'] == 'yes'

In [89]:
unrepaired_damage = a.loc[unrepaired, 'price']

In [90]:
unrepaired_damage.mean()

3855.3734782608694

In [91]:
repaired = a['unrepaired_damage'] == 'no'

In [92]:
repaired_damage = a.loc[repaired, 'price']

In [93]:
repaired_damage.mean()

7892.6990918264382

This is very first project on data science, i tried to give 100% afford on each and every corner of this project.
I know its not perfect and if you think that, is there anything on anywhere i need to improve or change please help me by those idea or method, any help or contribution will be highly appreciated and i will be very grateful to you.

Thank you for your time